Xamarin.Forms.Core/Device.cs

```diff
@@ -104,6 +104,65 @@ public static void BeginInvokeOnMainThread(Action action)
 			PlatformServices.BeginInvokeOnMainThread(action);
 		}
 
+		public static Task<T> InvokeOnMainThreadAsync<T>(Func<T> func)
+		{
+			var tcs = new TaskCompletionSource<T>();
+			BeginInvokeOnMainThread(() =>
+			{
+				try
+				{
+					var result = func();
+					tcs.SetResult(result);
+				}
+				catch (Exception ex)
+				{
+					tcs.SetException(ex);
+				}
+			});
+			return tcs.Task;
+		}
+
+		public static Task InvokeOnMainThreadAsync(Action action)
+		{
+			Func<object> dummyFunc = () => { action(); return null; };
+			return InvokeOnMainThreadAsync(dummyFunc);
+		}
+
+		public static Task<T> InvokeOnMainThreadAsync<T>(Func<Task<T>> funcTask)
+		{
+			var tcs = new TaskCompletionSource<T>();
+			BeginInvokeOnMainThread(
+				async() =>
+				{
+					try
+					{
+						var ret = await funcTask().ConfigureAwait(false);
+						tcs.SetResult(ret);
+					}
+					catch (Exception e)
+					{
+						tcs.SetException(e);
+					}
+				}
+			);
+
+			return tcs.Task;
+		}
+
+		public static Task InvokeOnMainThreadAsync(Func<Task> funcTask)
+		{
+			return InvokeOnMainThreadAsync(funcTask);
```

Comment:
This calls itself. Do something like this:

```cs
Func<Task<object>> dummyFunc = () => { funcTask(); return null; };
return InvokeOnMainThreadAsync(dummyFunc);
```

Reply:
good catch

Reply:
Thanks, fixed.